presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java

```diff
@@ -204,7 +204,14 @@ private SqlQueryExecution(
                     Optional.of(queryExplainer),
                     preparedQuery.getParameters(),
                     warningCollector);
-            this.analysis = analyzer.analyze(preparedQuery.getStatement());
+
+            try {
+                this.analysis = analyzer.analyze(preparedQuery.getStatement());
+            }
+            catch (RuntimeException e) {
+                stateMachine.getSession().getTransactionId().ifPresent(transactionManager::trySetInactive);
```

Comment:
Better fail it immediately, no need to wait for the transaction garbage collector

Reply:
Immediate fail added, however, the comment in TransactionManager::fail() suggests that cleanup is done later.

Reply:
I think we should do the similar check as it is here: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java#L759